deep_qa/training/models.py

```diff
@@ -28,6 +35,51 @@ def _summary_with_mask_info(self):
         flattened_layers = getattr(self, 'flattened_layers', self.layers)
         print_summary_with_masking(flattened_layers, getattr(self, 'container_nodes', None))
 
+    @overrides
+    def compile(self, params: Params):
+
+        """
+        The only reason we are overriding this method is because keras automatically wraps
+        our tensorflow optimiser in a keras wrapper, which we don't want. We override the
+        only method in Model which uses this attribute, ``_make_train_function``, which
+        raises an error if compile is not called first.
+        """
+        optimizer = params.get('optimizer')
+        self.gradient_clipping = params.pop("gradient_clipping", None)
+        super(DeepQaModel, self).compile(**params.as_dict())
+        self.optimizer = optimizer
+
+    @overrides
+    def _make_train_function(self):
+        """
+        We override this method so that we can use tensorflow optimisers directly.
+        This is desirable as tensorflow handles gradients of sparse tensors efficiently.
+        """
+        if not hasattr(self, 'train_function'):
+            raise RuntimeError('You must compile your model before using it.')
+        if self.train_function is None:
+            inputs = self._feed_inputs + self._feed_targets + self._feed_sample_weights
+            if self.uses_learning_phase and not isinstance(K.learning_phase(), int):
+                inputs += [K.learning_phase()]
+
+            # Here we override Keras to use tensorflow optimizers directly.
+            self.global_step = K.variable(0., name='global_step')
+            grads = self.optimizer.compute_gradients(self.total_loss, self._collected_trainable_weights)
+            if self.gradient_clipping is not None:
+                clip_type = self.gradient_clipping.pop("type")
+                clip_value = self.gradient_clipping.pop("value")
+                if clip_type == 'clip_by_norm':
+                    grads, _ = tensorflow.clip_by_global_norm(grads, clip_value)
```

Comment:
Ah thanks a lot, i'd always used tf.gradients and I hadn't realised the signatures would be different. Looks like the `None` problem has been fixed in TF 1.0 too 👍